pkg/controller/controller_test.go

```diff
@@ -2456,11 +2459,47 @@ func assertNumEvents(t *testing.T, strings []string, number int) {
 	}
 }
 
+// failfFunc is a type that defines the common signatures of T.Fatalf and
+// T.Errorf.
+type failfFunc func(t *testing.T, msg string, args ...interface{})
+
+func fatalf(t *testing.T, msg string, args ...interface{}) {
+	t.Log(string(debug.Stack()))
+	t.Fatalf(msg, args...)
```

Comment:
@krancour it does require some reading, but the resulting tests are much, much clearer to read.  [Example](https://github.com/pmorie/service-catalog/blob/38168abdf60b5e6edef8d1565122244e3afd5b01/pkg/controller/controller_test.go#L351) from #739:

```go
func TestReconcileBroker(t *testing.T) {
	fakeKubeClient, fakeCatalogClient, fakeBrokerClient, testController, _ := newTestController(t)
	fakeBrokerClient.CatalogClient.RetCatalog = getTestCatalog()

	testController.reconcileBroker(getTestBroker())

	actions := fakeCatalogClient.Actions()
	if err := assertNumberOfActions(t, actions, 2); err != nil {
		t.Fatal(err)
	}

	// first action should be a create action for a service class
	if _, err := assertCreate(actions[0], getTestServiceClass()); err != nil {
		t.Fatal(err)
	}

	// second action should be an update action for broker status subresource
	updatedBroker, err := assertUpdateStatus(actions[1], getTestBroker())
	if err != nil {
		t.Fatal(err)
	}
	if err := assertBrokerReadyTrue(updatedBroker); err != nil {
		t.Fatal(err)
	}

	// verify no kube resources created
	if err := assertNumberOfActions(t, fakeKubeClient.Actions(), 0); err != nil {
		t.Fatal(err)
	}

	events := getRecordedEvents(testController)
	if err := assertNumEvents(events, 1); err != nil {
		t.Fatal(err)
	}

	expectedEvent := api.EventTypeNormal + " " + successFetchedCatalogReason + " " + successFetchedCatalogMessage
	if e, a := expectedEvent, events[0]; e != a {
		t.Fatalf("Received unexpected event: %v", a)
	}
}
```

consider that versus:

```go
func TestReconcileBroker(t *testing.T) {
	fakeKubeClient, fakeCatalogClient, fakeBrokerClient, testController, _ := newTestController(t)

	fakeBrokerClient.CatalogClient.RetCatalog = getTestCatalog()

	testController.reconcileBroker(getTestBroker())

	actions := fakeCatalogClient.Actions()
	assertNumberOfActions(t, actions, 2)

	// first action should be a create action for a service class
	assertCreate(t, actions[0], getTestServiceClass())

	// second action should be an update action for broker status subresource
	updatedBroker := assertUpdateStatus(t, actions[1], getTestBroker())
	assertBrokerReadyTrue(t, updatedBroker)

	// verify no kube resources created
	assertNumberOfActions(t, fakeKubeClient.Actions(), 0)

	events := getRecordedEvents(testController)
	assertNumEvents(t, events, 1)

	expectedEvent := api.EventTypeNormal + " " + successFetchedCatalogReason + " " + successFetchedCatalogMessage
	if e, a := expectedEvent, events[0]; e != a {
		t.Fatalf("Received unexpected event: %v", a)
	}
}
```